### DISCUSSION TITLE
Kong golang custom plugin http2 & gRPC traffic capture

### DISCUSSION THREAD

**Discussion by prasadl02** [Kong/kong#13822] [Help] [answered]
Hello , 
I have implemented a custom plugin in golang which is intended to mirror all the requests and responses and sent to a remote server. I am facing an issue with http2 & gRPC upstreams. 

The request data is captured but while I receive the response from upstream server , the plugin is unable to mirror the response.

Below are the logs of kong docker. I am using the 3.8 version.

How do i solve this issue?

2024/11/02 15:12:35 [info] 2536#0: *638703 [kong] pb_rpc.lua:258  ------------- Request Data Captured --------------, client: 172.19.0.1
2024/11/02 15:12:35 [notice] 2536#0: *638703 [lua] init.lua:1358: access(): response buffering was turned off: incompatible HTTP version (2), client: 172.19.0.1
172.19.0.1 - - [02/Nov/2024:15:12:35 +0000] "POST /collector.CollectorProxy/CollectData HTTP/2.0" 200 7 "-" "grpc-node-js/1.8.13" kong_request_id: "ac4fc504fc716c65874e0b679b2aba73"

**Comment by xianghai2** [comment]
Hi, 
could you offer a simplified configuration package including the custom plugin and upstream service example? and detail how to reproduce this step by step. thanks.

**Comment by xianghai2** [comment]
@prasadl02， Kong is based on Nginx, and Nginx does not support http2 proxy pass to backend server. https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_http_version

**Comment by prasadl02** [comment]
No actually I am able to setup a service which proxies my request to a http2 server. The problem is that the custom plugin I developed in golang is using the response phase to get the response details. This is where the plugin is failing to get the response details of http2 and gRPC requests.

**Comment by xianghai2** [comment]
make sure the request is proxed to http2 server via http1.1 or http2 protocol ? this may be observed from backend http2 server.
then you may need to turn on proxy_buffering. https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_buffering

**Comment by prasadl02** [comment]
How do I turn it on using go plugin ?

**Comment by xianghai2** [comment, marked as answer]
This could be turned on inside Nginx conf.
